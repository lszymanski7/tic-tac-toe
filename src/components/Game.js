/* eslint-disable react-hooks/exhaustive-deps */

import React, { useState } from 'react'
import Board from './Board'
import Header from './Header'
import Score from './Score'
import checkWinner from './../utils/game-rules'

const Game = () => {
    // Default values for game states
    const defaultBoard = Array(9).fill(null)
    const defaultScore = {
        player1: 0,
        draw: 0,
        player2: 0
    }
    const defaultPlayers = {
        player1: {
            priority: true,
            turn: true,
            symbol: 'X'
        },
        player2: {
            priority: false,
            turn: false,
            symbol: 'O'
        }
    }
    const defaultWinner = null

    // Game states
    const [board, setBoard] = useState(defaultBoard)
    const [players, setPlayers] = useState(defaultPlayers)
    const [score, setScore] = useState(defaultScore)
    const [winner, setWinner] = useState(defaultWinner)

    // Update the board (change the value for a specific index)
    const updateBoard = (i) => {
        const nextState = board.map((element, index) => {
            if (index === i) {
                return players.player1.turn ? players.player1.symbol : players.player2.symbol
            } else {
                return element
            }
        })
        setBoard(nextState)

        return nextState
    }

    // Update the score (increase the value by 1)
    const updateScore = (winner) => {
        if (winner.symbol === 'X') {
            setScore({ ...score, player1: score.player1 + 1 })
        } else if (winner.symbol === 'O') {
            setScore({ ...score, player2: score.player2 + 1 })
        } else {
            setScore({ ...score, draw: score.draw + 1 })
        }
    }

    // Update the player's priority (determine who will start the game)
    const updatePriority = () => {
        if (players.player1.priority) {
            setPlayers({
                player1: { ...players.player1, priority: false, turn: false },
                player2: { ...players.player2, priority: true, turn: true }
            })
        } else {
            setPlayers({
                player1: { ...players.player1, priority: true, turn: true },
                player2: { ...players.player2, priority: false, turn: false }
            })
        }
    }

    // Update the player's turn (determine who has the next move)
    const updateTurn = () => {
        if (players.player1.turn) {
            setPlayers({
                player1: { ...players.player1, turn: false },
                player2: { ...players.player2, turn: true }
            })
        } else {
            setPlayers({
                player1: { ...players.player1, turn: true },
                player2: { ...players.player2, turn: false }
            })
        }
    }

    // Update the winner (determine who won the game)
    const updateWinner = (board) => {
        const nextState = checkWinner(board)
        setWinner(nextState)

        return nextState
    }

    // Reset the game
    const resetGame = () => {
        setBoard(defaultBoard)
        setWinner(defaultWinner)
        updatePriority()
    }

    // Call functions after clicking a cell on the board
    const handleOnClick = (i) => {
        if (winner === null && board[i] === null) {
            const updatedBoard = updateBoard(i)
            const updatedWinner = updateWinner(updatedBoard)
            updatedWinner === null ? updateTurn() : updateScore(updatedWinner)
        } 
        
        winner !== null && resetGame()
    }

    // Log values
    console.log('Board:', board)
    console.log('Score:', score)
    console.log('Players:', players)
    console.log('Winner:', winner)

    return (
        <div className="flexbox-column fullscreen">
            <Header />
            <Board
                board={board}
                handleOnClick={handleOnClick}
            />
            <Score
                players={players}
                score={score}
                winner={winner}
            />
        </div>
    )
}

export { Game as default }
