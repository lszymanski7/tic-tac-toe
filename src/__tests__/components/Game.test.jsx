import React from 'react'
import { render } from '@testing-library/react'
import Game from '../../components/Game'
import { ThemeProvider } from '../../contexts/ThemeContext'

describe('Game', () => {
  it('should be rendered correctly.', () => {
    const { container } = render(<Game />, { wrapper: ThemeProvider })
    expect(container).toMatchSnapshot()
  })
})
