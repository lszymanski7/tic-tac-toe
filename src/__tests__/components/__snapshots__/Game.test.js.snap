// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should be rendered correctly. 1`] = `
<div>
  <div
    class="fullscreen"
  >
    <header
      class="header"
    >
      <div
        class="header__widgets-container"
      >
        <a
          href="https://github.com/lszymanski7/tic-tac-toe"
        >
          <img
            alt="GitHub Logo"
            class="header__github-logo header__github-logo--light"
            src="github.svg"
          />
        </a>
        <div
          class="toggle-switch"
          data-testid="theme-switch"
        >
          <input
            class="toggle-switch__input"
            id="theme-switch"
            type="checkbox"
          />
          <label
            class="toggle-switch__label"
            for="theme-switch"
          >
            <span
              class="toggle-switch__icon moon-icon"
              data-testid="moon"
            />
            <span
              class="toggle-switch__icon sun-icon"
              data-testid="sun"
            />
            <span
              class="toggle-switch__slider"
            />
          </label>
        </div>
      </div>
      <h1
        class="header__game-name"
      >
        Tic-Tac-Toe
      </h1>
    </header>
    <div
      class="board"
    >
      <div
        class="board__row"
      >
        <div
          class="square square--light"
        >
          <button
            aria-label="Square 0"
            class="square__button"
          />
        </div>
        <div
          class="square square--light right left"
        >
          <button
            aria-label="Square 1"
            class="square__button"
          />
        </div>
        <div
          class="square square--light"
        >
          <button
            aria-label="Square 2"
            class="square__button"
          />
        </div>
      </div>
      <div
        class="board__row"
      >
        <div
          class="square square--light top bottom"
        >
          <button
            aria-label="Square 3"
            class="square__button"
          />
        </div>
        <div
          class="square square--light top right bottom left"
        >
          <button
            aria-label="Square 4"
            class="square__button"
          />
        </div>
        <div
          class="square square--light top bottom"
        >
          <button
            aria-label="Square 5"
            class="square__button"
          />
        </div>
      </div>
      <div
        class="board__row"
      >
        <div
          class="square square--light"
        >
          <button
            aria-label="Square 6"
            class="square__button"
          />
        </div>
        <div
          class="square square--light right left"
        >
          <button
            aria-label="Square 7"
            class="square__button"
          />
        </div>
        <div
          class="square square--light"
        >
          <button
            aria-label="Square 8"
            class="square__button"
          />
        </div>
      </div>
    </div>
    <div
      class="score"
    >
      <div
        class="score__container"
      >
        <p
          class="score__turn"
        >
          Player 1  [X]
        </p>
        <p>
          0
        </p>
      </div>
      <div
        class="score__container"
      >
        <p>
          Draw
        </p>
        <p>
          0
        </p>
      </div>
      <div
        class="score__container"
      >
        <p
          class=""
        >
          Player 2  [O]
        </p>
        <p>
          0
        </p>
      </div>
    </div>
  </div>
</div>
`;
